src/lib/server/links/utils.ts

```diff
@@ -1,15 +1,70 @@
 import QRCode from 'qrcode';
-import { LINKS } from './data';
+import type { Link, ParsedLink } from './types';
 
-export function parseLinkId(link: string): keyof typeof LINKS | undefined {
-  const cleanedLinkName = link
+export function parseLink(url: string): ParsedLink | undefined {
+  const hashBeginIdx = url.lastIndexOf('#');
+  let hashStr = '';
+  if (hashBeginIdx !== -1) {
+    hashStr = url.slice(hashBeginIdx + 1);
+    url = url.slice(0, hashBeginIdx);
+  }
+
+  const queryBeginIdx = url.indexOf('?');
+  let queryStr = '';
+  if (queryBeginIdx !== -1) {
+    queryStr = queryBeginIdx === -1 ? '' : url.slice(queryBeginIdx + 1);
+    url = url.slice(0, queryBeginIdx);
+  }
+
+  const relativePathnameStr = url
     .replace(/^\/+|\/+$/g, '')
     .replace(/[^a-zA-Z0-9-/]/g, '')
     .toLowerCase();
 
-  if (!(cleanedLinkName in LINKS)) return;
+  return { relativePathname: relativePathnameStr, query: queryStr, hash: hashStr };
+}
+
+export function findLinkId<ID extends string>(
+  segments: string[],
+  subCollection: Record<ID, string>,
+  separator = '/'
+): ID | undefined {
+  let id = segments.join(separator) as ID;
+  while (id.length > 0) {
+    if (subCollection[id]) return id;
+    id = id.slice(0, id.lastIndexOf(separator)) as ID;
+  }
+}
+
+/**
+ * Converts a URL to a link object.
+ * @param url The URL to convert.
+ * @param subCollection The collection of links to search.
+ * @returns The link object.
+ */
+export function parseLinkId<ID extends string>(
+  url: string,
+  subCollection: Record<ID, string>,
+  separator = '/'
+): Link<ID> | undefined {
+  const link = parseLink(url);
+  if (!link) return;
```

Comment:
can we unit test this condition?

Reply:
Attempted to resolve in <https://github.com/EthanThatOneKid/acmcsuf.com/pull/607/commits/564efa820265f2df4fffa3ba7651ef6beb1d6e87>.